lib/rules/no-undef-directives.js

```diff
@@ -0,0 +1,260 @@
+/**
+ * @author rzzf
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const utils = require('../utils')
+const casing = require('../utils/casing')
+
+/**
+ * @param {string} str
+ * @returns {string}
+ */
+function camelize(str) {
+  return str.replace(/-(\w)/g, (_, c) => (c ? c.toUpperCase() : ''))
+}
+
+/**
+ * @param {ObjectExpression} componentObject
+ * @returns { { node: Property, name: string }[] } Array of ASTNodes
+ */
+function getRegisteredDirectives(componentObject) {
+  const directivesNode = componentObject.properties.find(
+    (p) =>
+      p.type === 'Property' &&
+      utils.getStaticPropertyName(p) === 'directives' &&
+      p.value.type === 'ObjectExpression'
+  )
+
+  if (
+    !directivesNode ||
+    directivesNode.type !== 'Property' ||
+    directivesNode.value.type !== 'ObjectExpression'
+  ) {
+    return []
+  }
+
+  return directivesNode.value.properties
+    .filter((node) => node.type === 'Property')
+    .map((node) => {
+      const name = utils.getStaticPropertyName(node)
+      return name ? { node, name } : null
+    })
+    .filter((res) => !!res)
+}
+
+class DefinedInSetupDirectives {
+  constructor() {
+    /**
+     * Directive names
+     * @type {Set<string>}
+     */
+    this.names = new Set()
+  }
+
+  /**
+   * @param {string[]} names
+   */
+  addName(...names) {
+    for (const name of names) {
+      this.names.add(name)
+    }
+  }
+
+  /**
+   * @param {string} rawName
+   */
+  isDefinedDirective(rawName) {
+    const camelName = camelize(rawName)
+    const variableName = `v${casing.capitalize(camelName)}`
+    return this.names.has(variableName)
+  }
+}
+
+class DefinedInOptionDirectives {
+  constructor() {
+    /**
+     * Directive names
+     * @type {Set<string>}
+     */
+    this.names = new Set()
+  }
+
+  /**
+   * @param {string[]} names
+   */
+  addName(...names) {
+    for (const name of names) {
+      this.names.add(name)
+    }
+  }
```

Comment:
This is never called with multiple names, so can be simplified:


```suggestion
  /**
   * @param {string} name
   */
  addName(name) {
    this.names.add(name)
  }
```